rllib/algorithms/appo/tf/appo_tf_learner.py

```diff
@@ -0,0 +1,261 @@
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Any, Mapping
+
+from ray.rllib.policy.sample_batch import SampleBatch, MultiAgentBatch
+from ray.rllib.algorithms.appo.tf.appo_tf_rl_module import OLD_ACTION_DIST_KEY
+from ray.rllib.algorithms.impala.tf.vtrace_tf_v2 import make_time_major, vtrace_tf2
+from ray.rllib.algorithms.impala.tf.impala_tf_learner import ImpalaHPs, ImpalaTfLearner
+from ray.rllib.core.rl_module.marl_module import ModuleID
+from ray.rllib.core.learner.tf.tf_learner import TfLearner
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.framework import try_import_tf
+from ray.rllib.utils.metrics import LEARNER_STATS_KEY, ALL_MODULES
+from ray.rllib.utils.typing import TensorType
+
+_, tf, _ = try_import_tf()
+
+
+LEARNER_RESULTS_KL_KEY = "mean_kl_loss"
+
+
+@dataclass
+class AppoHPs(ImpalaHPs):
+    """Hyper-parameters for IMPALA.
+
+    Attributes:
+        rollout_frag_or_episode_len: The length of a rollout fragment or episode.
+            Used when making SampleBatches time major for computing loss.
+        recurrent_seq_len: The length of a recurrent sequence. Used when making
+            SampleBatches time major for computing loss.
+        discount_factor: The discount factor to use for computing returns.
+        vtrace_clip_rho_threshold: The rho threshold to use for clipping the
+            importance weights.
+        vtrace_clip_pg_rho_threshold: The rho threshold to use for clipping the
+            importance weights when computing the policy_gradient loss.
+        vtrace_drop_last_ts: Whether to drop the last timestep when computing the loss.
+            This is useful for stabilizing the loss.
+            NOTE: This shouldn't be True when training on environments where the rewards
+            come at the end of the episode.
+        vf_loss_coeff: The amount to weight the value function loss by when computing
+            the total loss.
+        entropy_coeff: The amount to weight the average entropy of the actions in the
+            SampleBatch towards the total_loss for module updates. The higher this
+            coefficient, the more that the policy network will be encouraged to output
+            distributions with higher entropy/std deviation, which will encourage
+            greater exploration.
+        kl_target: The target kl divergence loss coefficient to use for the KL loss.
+        kl_coeff: The coefficient to weight the KL divergence between the old policy
+            and the target policy towards the total loss for module updates.
+        tau: The factor by which to update the target policy network towards
+                the current policy network. Can range between 0 and 1.
+
```

Comment:
Are these inherited? I feel like the right way for inheritance is to only write the docstring for the new stuff, and refer to the base class for other parameter documentations. I actually don't know how sphinix will inherit documentations of dataclasses. I'd be curious to know. But for now let's just keep the docstring match the parameters that immediately come after, not the ones we have in the base class. It will be easier to maintain this way.

Reply:
sphinx isn't inheriting the docstring for the base class, hence my reason for doing this.